### PR TITLE
Add BinaryOutput

### PR DESCRIPTION
**Pull request by edknv**
### Goals :soccer:
This PR introduces the following classes for the pytorch backend:
- ModelOutput
- BinaryOutput

### Testing Details :mag:

Coverage:
```
Name                                            Stmts   Miss  Cover
-------------------------------------------------------------------
merlin/models/torch/outputs/__init__.py             0      0   100%
merlin/models/torch/outputs/base.py                23      0   100%
merlin/models/torch/outputs/classification.py      16      0   100%
-------------------------------------------------------------------
TOTAL                                              39      0   100%
```

### CODE REVIEW

File: merlin/models/torch/outputs/base.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        metrics: Sequence[Metric] = (),
+        name: Optional[str] = None,
+        schema: Optional[ColumnSchema] = None,
+    ):
+        """Initializes a ModelOutput object."""
+        super().__init__(*module, name=name)
+
+        self.loss = loss
+        self.metrics = metrics
+        self._output_schema = None
```
Comment (marcromeyn) [inline merlin/models/torch/outputs/base.py:54]:
```suggestion
        self._output_schema: Schema = Schema()
```

File: merlin/models/torch/outputs/base.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+        Parameters
+        ----------
+        schema: ColumnSchema or None
+            The schema defining the column properties.
+        """
+        self._output_schema = Schema([schema])
+
+    @property
+    def output_schema(self):
```
Comment (marcromeyn) [inline merlin/models/torch/outputs/base.py:71]: It's probably better to remove this and just do `self.output_schema` in the constructor. This because torchscript doesn't allow `@property` and only allows for methods.

File: merlin/models/torch/outputs/base.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    def eval(self):
+        """Sets the module in evaluation mode.
+
+        Returns
+        -------
+        nn.Module
+            The module in evaluation mode.
+        """
+        # Reset target
+        self.target = torch.zeros(1, dtype=torch.float32)
```
Comment (marcromeyn) [inline merlin/models/torch/outputs/base.py:90]: What about having a `create_target_buffer` method that can be used in the constructor as well?

```python
def create_target_buffer(self):
   self.register_buffer("target", torch.zeros(1, dtype=torch.float32))
```

That way, we could allow for `ModelOutput` to have targets with different dtypes if needed.

File: merlin/models/torch/outputs/classification.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    def __init__(
+        self,
+        loss=nn.BCEWithLogitsLoss(),
+        metrics: Sequence[Metric] = (
+            Accuracy(task="binary"),
+            AUROC(task="binary"),
+            Precision(task="binary"),
+            Recall(task="binary"),
+        ),
+        schema: Optional[ColumnSchema] = None,
```
Comment (marcromeyn) [inline merlin/models/torch/outputs/classification.py:48]: Maybe we should put schema on top, since that might be the argument that will be set the most.

File: merlin/models/torch/outputs/base.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+        self.loss = loss
+        self.metrics = metrics
+        self._output_schema = None
+
+        if schema:
+            self.setup_schema(schema)
+        self.register_buffer("target", torch.zeros(1, dtype=torch.float32))
+
+    def setup_schema(self, schema: Optional[ColumnSchema]):
```
Comment (marcromeyn) [inline merlin/models/torch/outputs/base.py:60]: Do you think we need to make things more explicit and turn this into `setup_column_schema`?

File: tests/unit/torch/outputs/test_classification.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+from torch import nn
+from torchmetrics import AUROC, Accuracy, Precision, Recall
+
+import merlin.dtypes as md
+import merlin.models.torch as mm
+from merlin.models.torch.utils import module_utils
+from merlin.schema import ColumnSchema, Schema
+
+
+class TestBinaryOutput:
```
Comment (marcromeyn) [inline tests/unit/torch/outputs/test_classification.py:26]: Would be great to add at least 2 extra tests:
1. Validate loss-calculation (after a forward-pass) 
2. Validate metric-calculation (after a forward-pass)

File: merlin/models/torch/outputs/base.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        The name of the model output.
+    schema: Optional[ColumnSchema]
+        The schema defining the column properties.
+    """
+
+    def __init__(
+        self,
+        *module: nn.Module,
+        loss: nn.Module,
+        metrics: Sequence[Metric] = (),
```
[the remark this answers is not in the document]
Comment (edknv) [reply merlin/models/torch/outputs/base.py:64]: `torcheval` looks interesting, but I think we should focus on just `torchmetrics` for now. `torcheval` says it's in alpha and is not stable, so I feel it could become an operational burden with maintaining dependencies, figuring out backward-incompatible changes, etc.

File: merlin/models/torch/outputs/base.py (hunk 5 above)
Comment (edknv) [reply merlin/models/torch/outputs/base.py:78]: I guess we could go either way, but I kept `setup_schema` since `output_schema` is a `Schema` not a `ColumnSchema`.

File: tests/unit/torch/outputs/test_classification.py (hunk 6 above)
Comment (edknv) [reply tests/unit/torch/outputs/test_classification.py:27]: I added a couple of tests that check the loss and the metrics, but it looks to me like the actual calculation would be better addressed after we implement `calculate_loss`.

**Review by marcromeyn — APPROVED** [review]
